plugins/logitech-scribe/fu-logitech-scribe-device.c

```diff
@@ -0,0 +1,603 @@
+				    &cmd_tmp,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	if (cmd_tmp != cmd) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "invalid upd message received, expected %x, got %x",
+			    cmd,
+			    cmd_tmp);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gchar *
+fu_logitech_scribe_device_compute_hash(GBytes *data)
+{
+	guint8 md5buf[HASH_VALUE_SIZE] = {0};
+	gsize data_len = sizeof(md5buf);
+	GChecksum *checksum = g_checksum_new(G_CHECKSUM_MD5);
+	g_checksum_update(checksum, g_bytes_get_data(data, NULL), g_bytes_get_size(data));
+	g_checksum_get_digest(checksum, (guint8 *)&md5buf, &data_len);
+	return g_base64_encode(md5buf, sizeof(md5buf));
+}
+
+static gboolean
+fu_logitech_scribe_device_query_data_size(FuLogitechScribeDevice *self,
+					  guchar unit_id,
+					  guchar control_selector,
+					  guint16 *data_size,
+					  GError **error)
+{
+	guint8 size_data[kDefaultUvcGetLenQueryControlSize] = {0x0};
+	struct uvc_xu_control_query size_query;
+	size_query.unit = unit_id;
+	size_query.selector = control_selector;
+	size_query.query = UVC_GET_LEN;
+	size_query.size = kDefaultUvcGetLenQueryControlSize;
+	size_query.data = size_data;
+
+	if (g_getenv("FWUPD_LOGITECH_SCRIBE_VERBOSE") != NULL) {
+		g_debug("Data size query request, unit: 0x%x selector: 0x%x",
+			(guchar)unit_id,
+			(guchar)control_selector);
+	}
+
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  UVCIOC_CTRL_QUERY,
+				  (guint8 *)&size_query,
+				  NULL,
+				  FU_LOGITECH_SCRIBE_DEVICE_IOCTL_TIMEOUT,
+				  error))
+		return FALSE;
+	// convert the data byte to int
+	*data_size = size_data[1] << 8 | size_data[0];
+	if (g_getenv("FWUPD_LOGITECH_SCRIBE_VERBOSE") != NULL) {
+		g_debug("Data size query response, size: %u unit: 0x%x selector: 0x%x",
+			*data_size,
+			(guchar)unit_id,
+			(guchar)control_selector);
+		fu_dump_raw(G_LOG_DOMAIN,
+			    "UVC_GET_LEN",
+			    size_data,
+			    kDefaultUvcGetLenQueryControlSize);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_scribe_device_get_xu_control(FuLogitechScribeDevice *self,
+					 guchar unit_id,
+					 guchar control_selector,
+					 guint16 data_size,
+					 guchar *data,
+					 GError **error)
+{
+	struct uvc_xu_control_query control_query;
+
+	if (g_getenv("FWUPD_LOGITECH_SCRIBE_VERBOSE") != NULL) {
+		g_debug("Get xu control request, size: %" G_GUINT16_FORMAT
+			" unit: 0x%x selector: 0x%x",
+			data_size,
+			(guchar)unit_id,
+			(guchar)control_selector);
+	}
+	control_query.unit = unit_id;
+	control_query.selector = control_selector;
+	control_query.query = UVC_GET_CUR;
+	control_query.size = data_size;
+	control_query.data = data;
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  UVCIOC_CTRL_QUERY,
+				  (guint8 *)&control_query,
+				  NULL,
+				  FU_LOGITECH_SCRIBE_DEVICE_IOCTL_TIMEOUT,
+				  error))
+		return FALSE;
+	for (guint i = 0; i < data_size; i++) {
+		data[i] = (guchar)data[i];
```

Comment:
eigh? what does this do?

Reply:
converting data to string. Any suggestion/recommendation here!

Reply:
`fu_strsafe()` certainly :)

Reply:
Fixed